book/forms.rst

```diff
@@ -621,6 +621,57 @@ large or whether you tried to submit text in a number field.
 
 .. _book-forms-type-reference:
 
+
+Groups based on a callable service
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you need some advanced logic to determine the validation groups that cannot
+be simply determined by a simple callback (for example that requires a few 
+services) you can use a service. Implement a service that implements __invoke
+which accepts a FormInterface as a parameter.
```

Comment:
to indicate that the next block is a PHP code block, you should end this sentence with `::` or add a new line with `.. code-block:: php`